src/boot/bootctl-install.c

```diff
@@ -423,12 +460,192 @@ static int copy_one_file(const char *esp_path, const char *name, bool force) {
         return ret;
 }
 
-static int install_binaries(const char *esp_path, const char *arch, bool force) {
+/* Parse shim's boot.csv https://github.com/rhboot/shim/blob/main/README.fallback
+ * Only parse file name and description in the first line. That's the only relevant information for
+ * systemd-boot. */
+static int parse_boot_csv_mem(
+                const char16_t *csv,
+                size_t size /* bytes */,
+                char **ret_loader_path,
+                char **ret_description) {
+
+        _cleanup_free_ char *buf = NULL, *loader_path = NULL, *description = NULL;
+        const char *s;
+        int r;
+
+        assert(csv);
+
+        /* Enforce unicode LE to avoid decoding garbage */
+        if (*csv != 0xfeff || size < 2)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid boot csv: %hx", *csv);
+
+        buf = utf16_to_utf8(csv+1, size-2);
+        if (!buf)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to decode boot csv. Must be UTF-16");
+
+        /* Only care about first line */
+        s = truncate_nl(buf);
+
+        r = extract_many_words(&s, ",", EXTRACT_KEEP_QUOTE, &loader_path, &description);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse boot csv: %m");
+        if (r < 2)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Missing file name and description in boot csv");
+
+        /* Shortest possible file name x.efi has four chars */
+        if (strlen(loader_path) < 4)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid file name in boot csv");
+        if (isempty(description))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Missing description in boot csv");
+
+        if (ret_loader_path)
+                *ret_loader_path = TAKE_PTR(loader_path);
+
+        if (ret_description)
+                *ret_description = TAKE_PTR(description);
+
+        return 0;
+}
+
+struct CopyLoaderArgs {
+        const char *source_path;
+        const char *esp_path;
+        bool force;
+        char **loader_path;
+        char **loader_description;
+};
+
+static int copy_loader(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
+        _cleanup_free_ char *source_path = NULL, *dest_path = NULL;
+        struct CopyLoaderArgs *args = ASSERT_PTR(userdata);
+        int r;
+
+        assert(path);
+        assert(args->source_path);
+        assert(args->esp_path);
+        assert(args->loader_path);
+        assert(args->loader_description);
+
+        if (event != RECURSE_DIR_ENTRY)
+                return RECURSE_DIR_CONTINUE;
+
+        r = chase(path, args->esp_path, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS|CHASE_NONEXISTENT, &dest_path, NULL);
```

Comment:
the esp is untrusted territory. please pin the fd when chasing i.e. use chase_and_open() to get the target fd right-away